Userland/Libraries/LibJS/Runtime/ErrorTypes.h

```diff
@@ -120,6 +120,7 @@
     M(PrivateFieldSetAccessorWithoutSetter, "Cannot set private field '{}' as accessor without setter")                                 \
     M(PrivateFieldSetMethod, "Cannot set private method '{}'")                                                                          \
     M(PromiseExecutorNotAFunction, "Promise executor must be a function")                                                               \
+    M(PromiseResolvedWithItself, "Cannot resolve promise with itself")                                                                  \
```

Comment:
I don't think this is needed - the real shortcoming is that host-defined hooks should return errors if needed (which we can probably continue to ignore for a little longer), just like various other supposedly infallible functions from the spec. This would only be a temporary solution anyway as eventually the `TypeError` allocation below should be fallible as well.

Reply:
hmm we need to deal with the String ENOMEM error in any case, and doing it manually in these instances is rather ugly. 

Like in the Promise case we'd have:
```c++
auto self_resolution_error_string = String::from_utf8("Cannot resolve promise with itself"sv);

auto self_resolution_error = self_resolution_error_string.is_error()
    ? Value { InternalError::create(realm, vm.error_message(VM::ErrorMessage::OutOfMemory)) }
    : Value { TypeError::create(realm, self_resolution_error_string.release_value()) };
```

Reply:
...unless you're cool with a `release_value_but_fixme` here in its stead (which is fine, I think my brain was actively ignoring that possibility here)

Reply:
What's wrong with `auto self_resolution_error = MUST_OR_THROW_OOM(vm, TypeError::create(realm, "Cannot resolve promise with itself"))`? The `PromiseResolvingFunction` should be able to handle that just fine:

https://github.com/trflynn89/serenity/blob/87b544aba3932f389b7968dceb4c9fd85ca9070d/Userland/Libraries/LibJS/Runtime/PromiseResolvingFunction.cpp#L35-L38

Reply:
...this is different from your approach, but I wouldn't even try to continue to reject the promise in an OOM state - we could bail out early and requeue OOM'd promises for processing at a later time.

Reply:
> What's wrong with `auto self_resolution_error = MUST_OR_THROW_OOM(vm, TypeError::create(realm, "Cannot resolve promise with itself"))`? The `PromiseResolvingFunction` should be able to handle that just fine:

:man_facepalming: When I first looked at this function, I thought it had a much larger footprint. Turns out it's *already* expected to return a ThrowCompletionOr:
https://github.com/SerenityOS/serenity/blob/2692db869963ed10947b05f13eade5fcbd5951dc/Userland/Libraries/LibJS/Runtime/PromiseResolvingFunction.h#L29